### PR TITLE
Fix EXPORT_ES6 + USE_PTHREADS

### PR DESCRIPTION
**Pull request by VirtualTim**
I haven't managed to get this working, and I think I'm stuck, but I figured I'd create this PR so people can pick up where I left off. Or point out if I'm doing something wrong.

### CODE REVIEW

**Comment by VirtualTim** [comment]
Currently "_emscripten_futex_wake" is being optimized away, so I need to compile with `-s EXPORTED_FUNCTIONS='["_emscripten_futex_wake"]'`. I'm sure there is a way to force-preserve this, but I don't know where to add that. Anyone want to point me to the right file to add this to?

File: emcc.py [hunk 1]
```diff
@@ -2771,14 +2771,19 @@ def modularize():
       # after document.currentScript is gone, so we save it.
       # (when MODULARIZE_INSTANCE, an instance is created
       # immediately anyhow, like in non-modularize mode)
+      if shared.Settings.EXPORT_ES6:
+        GETSCRIPT = "performance.getEntries().slice(-1)[0].name"
```
Comment (curiousdannii) [inline emcc.py]: If this returns the current script's path, cool, good to know there's another method, but wouldn't `import.meta.url` be simpler?

Comment (VirtualTim) [reply emcc.py]: Huh, I didn't know that was a thing. Yeah that's much better.

**Comment by VirtualTim** [comment]
Things seem to be mostly working now. I don't know why the test are failing, those webgl tests seem kind of flaky on firefox, so maybe it's unrelated.
It doesn't yet work with `WASM=1`, it throws with `TypeError: Cannot assign to read only property` trying to reassign functions.

**Comment by VirtualTim** [comment]
@kripken or @sbc100, since you both commented on the original issue, could one of you take a look at this PR? I think I've got it working, I'm pretty sure the test failures are unrelated.

I also want to add a test case for this, but I'm not really sure how to test something like this. I think I need some JS which loads the compiled ES6 module, but I don't rally know how to integrate this with the existing test framework.

File: emcc.py [hunk 2]
```diff
@@ -2771,14 +2771,19 @@ def modularize():
       # after document.currentScript is gone, so we save it.
       # (when MODULARIZE_INSTANCE, an instance is created
       # immediately anyhow, like in non-modularize mode)
+      if shared.Settings.EXPORT_ES6:
+        GETSCRIPT = "import.meta.url"
+      else:
+        GETSCRIPT = "typeof document !== 'undefined' && document.currentScript ? document.currentScript.src : undefined"
```
Comment (sbc100) [inline emcc.py]: No need to make this local variable uppercase.

Would this be better names "script_url" or something like that?

Why is the js var called `_scripteDir`?   Isn't it the full URL?

File: emcc.py [hunk 3]
```diff
@@ -2771,14 +2771,19 @@ def modularize():
...
       src = '''
 var %(EXPORT_NAME)s = (function() {
-  var _scriptDir = typeof document !== 'undefined' && document.currentScript ? document.currentScript.src : undefined;
+  var _scriptDir = %(GETSCRIPT)s;
```
Comment (sbc100) [inline emcc.py]: No need to uppercase here either.  EXPORT_NAME is uppercase because it corresponds to a Settings member.

File: emscripten.py [hunk 4]
```diff
@@ -1666,6 +1666,10 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
 
   shared.Settings.MODULE_EXPORTS = module_exports = exported_implemented_functions + function_tables(function_table_data)
 
+  # Should only need to do this in the main thread.
+  if shared.Settings.USE_PTHREADS:
+    receiving = "if (!ENVIRONMENT_IS_PTHREAD) {\n" + receiving + "}\n"
```
Comment (sbc100) [inline emscripten.py]: But the stuff added to receiving below we do want on non-main-threads?  Can you explain, perhaps write more in the comment here?

File: src/shell_pthreads.js [hunk 5]
```diff
@@ -4,10 +4,10 @@
 // 3) We could be an application pthread running in a worker. (ENVIRONMENT_IS_WORKER == true and ENVIRONMENT_IS_PTHREAD == true)
 #if USE_PTHREADS
 
-if (typeof ENVIRONMENT_IS_PTHREAD === 'undefined') {
+var ENVIRONMENT_IS_PTHREAD = typeof importScripts === 'function';
```
Comment (sbc100) [inline src/shell_pthreads.js]: This seems to contradict the commend above.   Not all workers are pthreads, but this change looks like it would set ENVIRONMENT_IS_PTHREAD for all workers.

File: emscripten.py (hunk 4 above)
Comment (VirtualTim) [reply emscripten.py]: In my testing it seems like these functions were already shared with the non-main-threads by the main thread. Calling these from a non-main-thread attempted to redefine them (as the same thing). No idea if that's how it's supposed to work.

File: src/shell_pthreads.js (hunk 5 above)
Comment (VirtualTim) [reply src/shell_pthreads.js]: Crap you're right. I'll have to look at another way of determining if the script is being called from a thread.

File: emcc.py (hunk 2 above)
Comment (VirtualTim) [reply emcc.py]: Yeah good point. I didn't name that, but it is a bit confusing. `_scriptDir` can be a blob, or path to a script. Later on `_scriptDir` is assigned to `scriptDirectory`, which contains the directory of the script. I'm not sure what `scriptDirectory` is used for. 
Anyway `import.meta.url` seems to return the same thing as `document.currentScript.src`, the difference being that `import.meta.url` works in a module, while `document.currentScript.src` doesn't.

File: emcc.py (hunk 3 above)
Comment (VirtualTim) [reply emcc.py]: Fair enough. I'll change this.

File: src/shell_pthreads.js (hunk 5 above)
Comment (VirtualTim) [reply src/shell_pthreads.js]: I've changed what I'm doing here, instead of trying to determine if we're in a thread from a script, I'm attaching ENVIRONMENT_IS_PTHREAD to the module that gets passed from the worker.
I'm pretty sure that's the right way to do this?

**Comment by VirtualTim** [comment]
Well the tests failed because the `test_openjpeg`  test didn't find the symbol `__memory_base`. Not sure why `test_the_bullet` failed, but that seems to be failing for everyone at the moment.
I'm pretty sure those are unrelated to my changes.

I'm happy for this to be merged, if everyone else is happy with the changes.

File: emcc.py [hunk 6]
```diff
@@ -2771,14 +2771,19 @@ def modularize():
       # after document.currentScript is gone, so we save it.
       # (when MODULARIZE_INSTANCE, an instance is created
       # immediately anyhow, like in non-modularize mode)
+      if shared.Settings.EXPORT_ES6:
+        script_url = "import.meta.url"
```
Comment (kripken) [inline emcc.py]: this can use a comment - i'm not sure what it means?

File: emscripten.py (hunk 4 above)
Comment (kripken) [reply emscripten.py]: Based on where this line is, it just puts the runtime assertions stuff behind this check. What's special about them? As @sbc100 said, I think this should all be needed in pthreads too - do you have a test that fails without this change?

File: src/worker.js [hunk 7]
```diff
@@ -23,7 +23,6 @@ var buffer; // All pthreads share the same Emscripten HEAP as SharedArrayBuffer
 var DYNAMICTOP_PTR = 0;
 var DYNAMIC_BASE = 0;
 
-var ENVIRONMENT_IS_PTHREAD = true;
```
Comment (kripken) [inline src/worker.js]: is this not needed? it's outside the MODULARIZE scope for that mode, which maybe does need this?

**Comment by kripken** [comment]
The `test_openjpeg` failures are unrelated, yeah, that was an issue we had (that should now be fixed on incoming I think).

File: emcc.py (hunk 6 above)
Comment (VirtualTim) [reply emcc.py]: It's basically the same as `document.currentScript.src`, but for modules. `document.currentScript.src` is not available to modules (see: https://html.spec.whatwg.org/multipage/dom.html#dom-document-currentscript), so `import.meta.url` (or some other way) has to be used instead.

Basically the result is the same, just one works for modules and the other doesn't.
Do you still think this needs a comment?

File: emscripten.py (hunk 4 above)
Comment (VirtualTim) [reply emscripten.py]: When I tried to run the code before this change I was getting errors about the vars in this block being redefined. It seems like when initialising a thread these functions have already been defined (not sure how). Normally var redefinition isn't an issue as everything gets hoisted, so I assumed it was an ES6 module thing. I'll have to do some more testing, I can't remember exactly what error I was trying to avoid.

Although looking at that again I should add a check for if `receiving` is an empty string.

File: src/worker.js (hunk 7 above)
Comment (VirtualTim) [reply src/worker.js]: Instead of just having it as a var, I thought the best way to handle this was to set `Module['ENVIRONMENT_IS_PTHREAD']` in the thread, and in `shell_pthreads.js` set `var ENVIRONMENT_IS_PTHREAD` based off `Module.ENVIRONMENT_IS_PTHREAD`. That way you don't get weird scoping issues, and `ENVIRONMENT_IS_PTHREAD` is explicitly set.
I think it makes sense to do this for all pthreads code, so it's not a module specific change.

File: emscripten.py (hunk 4 above)
Comment (kripken) [reply emscripten.py]: That's a very odd error - I don't think we use read-only properties anywhere?

File: emcc.py (hunk 6 above)
Comment (kripken) [reply emcc.py]: Thanks! Yeah, no need for a comment then.

File: src/worker.js (hunk 7 above)
Comment (kripken) [reply src/worker.js]: Oh, so something similar to the src/preamble.js change in https://github.com/emscripten-core/emscripten/pull/8365/files ? Yeah, I think that direction makes sense.

Comment (VirtualTim) [reply src/worker.js]: Yeah, but this case it's only passing a bool around, and only on initialisation, so I think overhead would be negligible.

File: emscripten.py (hunk 4 above)
Comment (VirtualTim) [reply emscripten.py]: Yeah my guess is that in ES6 modules properties are read-only by default. I don't see any way to change that without changing the way we define properties. That seems like a lot of work for very little gain.
I was hoping I could just change it to writable after creation, like `Object.defineProperty(asm, '__ZSt18uncaught_exceptionv', {configurable:true, writable:true});` but that just throws `Uncaught TypeError: Cannot redefine property: __ZSt18uncaught_exceptionv`.

The affect is pretty edge case, and the impact is that there's slightly less debugging statements. Personally I don't think this is that big of a deal, do you? This could always get merged and an issue opened to fix runtime assertions inside an ES6 module.

Comment (VirtualTim) [reply emscripten.py]: Actually after taking another look, the code looks like:

```
var real___ZSt18uncaught_exceptionv = asm["__ZSt18uncaught_exceptionv"];
asm["__ZSt18uncaught_exceptionv"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return real___ZSt18uncaught_exceptionv.apply(null, arguments);
};
```
...
```
var __ZSt18uncaught_exceptionv = Module["__ZSt18uncaught_exceptionv"] = function() {
  assert(runtimeInitialized, 'you need to wait for the runtime to be ready (e.g. wait for main() to be called)');
  assert(!runtimeExited, 'the runtime was exited (use NO_EXIT_RUNTIME to keep it alive after main() exits)');
  return asm["__ZSt18uncaught_exceptionv"].apply(null, arguments)
};
```

So if the code calls `Module.__ZSt18uncaught_exceptionv` it already goes through the runtime checks. So it seems like that entire block of code may be redundant in WASM anyway.

Comment (kripken) [reply emscripten.py]: I'm having a hard time following this - what is failing, and why? I don't see how to connect the error message from before about read-only properties to those code quotes.

Comment (VirtualTim) [reply emscripten.py]: So it fails on the line `asm["__ZSt18uncaught_exceptionv"] = function() {`. with `Cannot redefine property: __ZSt18uncaught_exceptionv`
I'm not sure where this property is initially defined, probably in the WASM?
The function is modified so that it goes through runtime assertions before being called. However when the function `Module["__ZSt18uncaught_exceptionv"]` is called it goes through the assertions, into `asm["__ZSt18uncaught_exceptionv"]`, then through the same assertions again. It seems a little redundant to me, doubling up the assertions.

So to sum up:
**The bug:**
The line `asm["__ZSt18uncaught_exceptionv"] = function() {`. throws with `Cannot redefine property: __ZSt18uncaught_exceptionv`

**My change:**
Don't redefine (to inject runtime assertions) functions when inside an ES6 module pthread.

**The impact:**
In an ES6 module pthread asm functions will only jump through one runtime assertion check, instead of the usual two.

If it would make it clearer I could post some example code, but you would either need to build it off my branch, or I could post the compiled example.

Comment (kripken) [reply emscripten.py]: Yeah, maybe post the compiled example, and instructions for how to run it. I've never seen that error before, I'm mystified.

Are you perhaps running this with `use strict` or some special JS VM flag?

Comment (VirtualTim) [reply emscripten.py]: I tried to attach a zip to this comment, but it didn't work. [Hopefully you can access this link?](https://hexmet-my.sharepoint.com/:u:/g/personal/tim_lander_hexagon_com/EX0XSdcY6GtDmtpM-7Eb16IBgn29uA-_ojG1F41NXzsRlw?e=vcB2WN)

It can only be run though Chrome with the "Experimental Web Platform features" flag enabled.
I ran the server like: `emrun --no_browser --no_emrun_detect --hostname $(hostname -I) --port 8080 module_test/test.html`, so nothing fancy

If you look at `line 3291` you'll see this change. If you change that line to `if (true) {` the code will throw with `TypeError: Cannot assign to read only property '__ZN14test_wrapper_212pthread_testEi' of object '[object Object]'`

Hopefully this makes things clearer.

Comment (kripken) [reply emscripten.py]: Thanks for the testcase - I'm not sure I understand the instructions after it, though. I set that flag, and then on line 3291 I see
```
var _setThrew = Module["_setThrew"] = function() {
```
and I changed that to
```
if (true) {
```
Now when I run it I get `you need to wait for the runtime to be ready`. So I don't see that "read only property" error?

Comment (VirtualTim) [reply emscripten.py]: Uh, I think you're looking at the wrong line. `var _setThrew = Module["_setThrew"] = function() {` is on line 3**92**1, I want you to look at line 3**29**1.

Line 3291 should have `if (!ENVIRONMENT_IS_PTHREAD) {`

Comment (kripken) [reply emscripten.py]: Thanks, and sorry for my mistake earlier!

Ok, what I think is happening here is: when loaded as an ES6 module, the wasm exports object is an ES6 module export object, or behaves like one, and those are more "static" than normal JS objects. And we try to replace them with wrappers to add extra runtime error handling, and that fails.

But that doesn't explain why this is a problem only on workers but not on the main thread. It sounds like that could be a browser bug, in fact. That asm object is a wasm export object on both locations, so it should behave the same? Unless someone only the worker is an ES6 context - I'm not too familiar with how those are created?

Comment (kripken) [reply emscripten.py]: Regardless, we can split this out from the rest of this PR, if it's separable.

Comment (VirtualTim) [reply emscripten.py]: Yeah I don't really know too much about all that stuff either.
The change on that line allows EXPORT_ES6 + USE_PTHREADS to work with ASSERTIONS != 0. I think the best thing is to merge this PR as is, since by default ASSERTIONS=1.

Perhaps after this is merged another bug can be opened to fix those assertions?\

If you'd prefer I'm happy to remove that line, and just say that currently EXPORT_ES6 + USE_PTHREADS only work with ASSERTIONS=0.

Comment (kripken) [reply emscripten.py]: Yeah, let's separate it out - a comment that says something like your last line there sounds good enough for now.

Comment (VirtualTim) [reply emscripten.py]: OK, I changed it to an assert, since the next line would just throw anyway. At least this way the developer gets a clear message about this not working. Once this is merged I'll open another issue to fix USE_PTHREADS + EXPORT_ES6 +ASSERTIONS.

**Comment by kripken** [comment]
Merging in latest incoming to here will hopefully fix those openjpeg tests and other errors.

**Comment by VirtualTim** [comment]
Well I merged everything from Incoming into my PR, but unfortunately I did a rebase instead of a regular merge. I'm not sure how to fix this without creating more mess (or if it even needs fixing?).
What do you think @kripken? Is this mergable, or should I just delete this PR and re-create it?
At least the tests all passed...

File: emscripten.py [hunk 8]
```diff
@@ -1699,6 +1699,10 @@ def create_receiving(function_table_data, function_tables_defs, exported_impleme
 
   shared.Settings.MODULE_EXPORTS = module_exports = exported_implemented_functions + function_tables(function_table_data)
 
+  # Should only need to do this in the main thread.
```
Comment (kripken) [inline emscripten.py]: Please clarify this in the comment. I believe the issue is that when assertions are on, we emit code that is incompatible with ES6 currently? Can add a TODO in the comment to fix that.

File: src/worker.js [hunk 9]
```diff
@@ -35,6 +34,10 @@ var __performance_now_clock_drift = 0;
 // Cannot use console.log or console.error in a web worker, since that would risk a browser deadlock! https://bugzilla.mozilla.org/show_bug.cgi?id=1049091
 // Therefore implement custom logging facility for threads running in a worker, which queue the messages to main thread to print.
 var Module = {};
+#if EXPORT_ES6
```
Comment (kripken) [inline src/worker.js]: i think we should also emit these lines if MODULARIZE as we do assign to them there as well (line 144/158), and for the same reason?

File: src/worker.js [hunk 10]
```diff
@@ -149,6 +163,7 @@ this.onmessage = function(e) {
       if (typeof FS !== 'undefined' && typeof FS.createStandardStreams === 'function') FS.createStandardStreams();
 #endif
       postMessage({ cmd: 'loaded' });
+#endif
```
Comment (kripken) [inline src/worker.js]: please add a comment `// MODULARIZE && EXPORT_ES6` so it's easier to match up the if and its end

File: src/worker.js (hunk 9 above)
Comment (VirtualTim) [reply src/worker.js]: The difference is that in normal (aka sloppy) JS you can define things without var/const/let. So defining `PThread` like `PThread = Module['PThread'];` is perfectly valid. However this is not valid inside an ES6 module as that enforces strict mode. That's why I had to declare them here.
I can change `#if EXPORT_ES6` to `#if MODULARIZE` if you want, but there may be a reason that it was initially done like that and I didn't want to break existing modularize code.

File: emscripten.py (hunk 8 above)
Comment (VirtualTim) [reply emscripten.py]: Good point, I missed the comment what I changed the behaviour.

**Comment by kripken** [comment]
Sounds good. I'll do the modularize changes as a separate commit later, which is probably better anyhow.

I'm rerunning the chrome tests which all failed here - probably that's just bad luck, and the rerun will pass. I'll land this if so.
